### ISSUE TITLE
NiftiMasker mask_strategy generate wrong mask

### ISSUE THREAD

**Issue by LeiGuo0812** [nilearn/nilearn#3358]
Hi, I'm new to nilearn and I am following the tutorial to use NiftiMasker. But when I tried different parameters of mask_strategy, the mask that "gm-template" and "wm-template" generated seemed wired.

Nilearn version:
0.8.0
### Expected behavior
generate gray matter and white matter mask

### Actual behavior
generated cerebellum mask and right brain white matter (partly)

### Steps and code to reproduce bug

```python
from nilearn.maskers import NiftiMasker
from nilearn.plotting import plot_roi
import numpy as np
import matplotlib.pyplot as plt
from nilearn import datasets

n_samples = 20

localizer_dataset = datasets.fetch_localizer_contrasts(
    ['left button press (auditory cue)'],
    n_subjects=n_samples, legacy_format=False
)

contrast_map_filenames = localizer_dataset.cmaps

# whole brain masker
mask_whole_brain = NiftiMasker(mask_strategy='whole-brain-template')
mask_whole_brain.fit(contrast_map_filenames)
plot_roi(mask_whole_brain.mask_img_, display_mode = 'z', cut_coords = 4)

# gray matter masker
mask_gm = NiftiMasker(mask_strategy='gm-template')
mask_gm.fit(contrast_map_filenames)
plot_roi(mask_gm.mask_img_, display_mode = 'z', cut_coords = 4)

# white matter masker
mask_wm = NiftiMasker(mask_strategy='wm-template')
mask_wm.fit(contrast_map_filenames)
plot_roi(mask_wm.mask_img_, display_mode = 'z', cut_coords = 4)
```
![gm-template](https://user-images.githubusercontent.com/50766698/188766359-bc2435b0-659e-4a29-a653-e804ef4a3d54.png)
![wm-template](https://user-images.githubusercontent.com/50766698/188766368-ac3778d6-8236-4760-91bb-1a0c53131d77.png)

**Comment by bthirion** [comment]
Thx for reporting.
This is a bad behavior indeed.  I think that this is because mask interpolation is not done correctly and results in very conservative maks estimate. 
I don't have time to dig into that now, hopefully later this week...
Meanwhile, I advise you to create the gm/wm mask you want by calling `nilearn.image.resample_to_img` and adjusting parameters; then you call NiftiMasker with a mask_img this is exactly what you want and does not need to be resampled.
HTH

**Comment by jeromedockes** [comment]
Even when resampling to the 2-mm MNI template the issue occurs:

```python
from nilearn import masking, datasets, plotting

mni = datasets.load_mni152_template(2)
mask = masking.compute_brain_mask(mni, mask_type="gm")
plotting.plot_roi(mask, display_mode="z", cut_coords=10)
```
![fig](https://user-images.githubusercontent.com/9196501/188873263-4390dfe8-2f47-4e77-8fcb-fda98d257be5.png)

The threshold that determines the mask can be chosen with `mask_args`; for example in the original script we get better results when using:

```
mask_gm = NiftiMasker(mask_strategy='gm-template', mask_args={"threshold": .2})
mask_wm = NiftiMasker(mask_strategy='wm-template', mask_args={"threshold": .2})
```
(the default threshold is `.5`)

I have tried to see if we can choose a default threshold in the call to
`compute_mask` ([here in the niftimasker](https://github.com/nilearn/nilearn/blob/9aa503f599f0835aea2bf9e2720bd5d3e33e941e/nilearn/maskers/nifti_masker.py#L34)),
but (i) changing it would change the output of existing code and (ii) I have the
impression we will have a hard time finding a threshold that works better than
the current one in all cases (though I think overall `.5` may be a bit too high
for the wm and gm mni templates in `nilearn.datasets`, even if it works well when
no resampling is necessary). Maybe we should try to highlight `mask_args` more 
in the docstring and point outthat often the threshold will need to be tuned for the
target image?

But otherwise I think it is easier to resample a mask rather than compute a new
one by resampling and thresholding a template; it works well and avoids having
to tune the threshold:

```python
mask_gm = NiftiMasker(mask_img=datasets.load_mni152_gm_mask())
mask_wm = NiftiMasker(mask_img=datasets.load_mni152_wm_mask())
```

![gm](https://user-images.githubusercontent.com/9196501/188873377-5ac5944f-cd9d-4722-9e9d-92d313db154b.png)
![wm](https://user-images.githubusercontent.com/9196501/188873380-723cfa8a-c7df-45bc-83d1-d9c4a62d1ac6.png)

**Comment by LeiGuo0812** [comment]
Thanks a lot for your reply!
I checked `nilearn.image.resample_to_img` function and found that it only resamples the resolution of one image to another. So I think maybe you are referring to the function `nilearn.masking.compute_brain_mask`. (Because this function can generate mask according to an image resolution and choose the tissue I want).

I found that the correctness of the generated mask seriously depends on the image provided to `nilearn.masking.compute_brain_mask`.  I used totally the same parameter except for the image in different resolutions. But it generated very different masks.

```python
# High resolution 1-1-1 mm ch2 brain
wm_mask = compute_brain_mask('./ch2.nii.gz', 
                             mask_type = 'wm',
                             threshold = 0.5)

# Low resolution 3-3-3 mm localizer image
wm_mask2 = compute_brain_mask(contrast_map_filenames[0], 
                             mask_type = 'wm',
                             threshold = 0.5)
```
High resolution
![wm_mask](https://user-images.githubusercontent.com/50766698/188873236-55a06ed8-5fe1-4755-ba43-692d729644c2.png)

Low resolution
![wm_mask2](https://user-images.githubusercontent.com/50766698/188873250-1181fa84-4ac2-4f5e-83a8-d36b817fcee8.png)

When `mask_type='gm'`, the masks generated from two different resolution images were also different. But it's fine for `mask_type='whole_brain'`. I am confusing because the function document said `Only the shape and affine of target_img will be used here.` so it should not affect the general shape of the mask.

**Comment by jeromedockes** [comment]
what `compute_brain_mask` does is:
- resample the template to the target image's resolution
- threshold it

so both the template (wm, gm, whole-brain) and the target image's resolution
(affine) influence what the resampled template looks like, and what is the
appropriate threshold. So it is possible that the threshold needs to be tuned
for a particular combination of template and target image. To avoid having to
deal with this, the easiest thing is to directly use a mask (or pass `mask_img`
to NiftiMasker), possibly resampling it to the target image's resolution by
using `resample_to_img(interpolation="nearest")` or by passing the target
image's affine and shape to the NiftiMasker constructor as `target_affine` and
`target_shape`

**Comment by bthirion** [comment]
The procedure includes the extraction of the largest connected component ... This would explain why we may loose an hemisphere ... add `connected=False` to get rid of it.

**Comment by LeiGuo0812** [comment]
Thanks a lot for your explanation and solutions. I think a better way is to use a pre-defined mask, as @jeromedockes recommended. 

Maybe the example here [](https://nilearn.github.io/stable/auto_examples/07_advanced/plot_localizer_mass_univariate_methods.html#sphx-glr-auto-examples-07-advanced-plot-localizer-mass-univariate-methods-py) could have an adjusting. I first followed this example and the mask generated might be undesired. 

Thanks again for your reply!

**Comment by bthirion** [comment]
Should we add a short caveat to the example then ?
@LeiGuo0812 are you willing to do this  ?
Best,

**Comment by LeiGuo0812** [comment]
Sure, I'm very glad to contribute! But I am very new to GitHub, and I'm learning the instructions in [http://nilearn.github.io/stable/development.html](http://nilearn.github.io/stable/development.html). This may take a few days. Thanks again!

**Comment by jeromedockes** [comment]
@LeiGuo0812  if you are having trouble with GitHub and how to open pull requests etc. let us know we are happy to help

**Comment by tejas-savalia** [comment]
Hello! I'm not sure if this is currently being worked on for the next update but just want to confirm that it's not fixed in version 0.10.4. 

Currently, using a 'gm-template' as an option to NiftiMasker, produces a much smaller mask and does not cover most of the gray matter. Reproducing the example above:

`from nilearn import datasets`

`n_samples = 20`

`localizer_dataset = datasets.fetch_localizer_contrasts(
    ['left button press (auditory cue)'],
    n_subjects=n_samples, legacy_format=False
)`

`contrast_map_filenames = localizer_dataset.cmaps`

`# gray matter masker`
`mask_gm = NiftiMasker(mask_strategy='gm-template')`
 #`plotting.view_img(mask_gm.inverse_transform(mask_gm.fit_transform(contrast_map_filenames)[0]))`

Produces:
![Image](https://github.com/user-attachments/assets/7b4cd68e-e239-4c5b-9378-f01f9e860b61)

Whereas, using the following
`mask_gm = NiftiMasker(mask_img=datasets.load_mni152_gm_mask(resolution=2))`
`plotting.view_img(mask_gm.inverse_transform(mask_gm.fit_transform(contrast_map_filenames)[0]))`

Produces the correct mask:
![Image](https://github.com/user-attachments/assets/c57507e8-9db9-4261-b965-b04cc9c38767)

**Comment by bthirion** [comment]
Oops, indeed.

**Comment by Remi-Gau** [comment]
> Thanks a lot for your explanation and solutions. I think a better way is to use a pre-defined mask, as [@jeromedockes](https://github.com/jeromedockes) recommended.
> 
> Maybe the example here [Massively univariate analysis of a motor task from the Localizer dataset](https://nilearn.github.io/stable/auto_examples/07_advanced/plot_localizer_mass_univariate_methods.html#sphx-glr-auto-examples-07-advanced-plot-localizer-mass-univariate-methods-py) could have an adjusting. I first followed this example and the mask generated might be undesired.

I am not sure this is the example to adjust as it does not use `mask_strategy`

**Comment by Remi-Gau** [comment]
> just want to confirm that it's not fixed in version 0.10.4.

from re-reading the whole thread it seems to me to be more of a feature, and that if one wants to restrict the analysis to grey / white matter it would be more relieable to either tune the strategy or better, pass a mask_img directly

from @jeromedockes 

> I have the impression we will have a hard time finding a threshold that works better than
the current one in all cases [...].
> Maybe we should try to highlight mask_args more
in the docstring and point outthat often the threshold will need to be tuned for the
target image?

**Comment by Remi-Gau** [comment]
If one example should be updated to mention the caveats and further options, it should be:
https://nilearn.github.io/dev/auto_examples/06_manipulating_images/plot_mask_computation.html#sphx-glr-auto-examples-06-manipulating-images-plot-mask-computation-py
